API example

yulai46 opened the pull request.

proofit404 commented:
Привет!

Предлагаю в существующем пакете django/example создать следующие подпакеты:
* `api` - файлы с `@model_view_set`, определёнными из `Injector`, ссылаются на те же самые stories из `implemented`
* `serializers` - набор стандартных сериалайзеров ничем не отличающихся от классического drf.
* `urls` - из файла превратить в подпакет. 
** `urls/__init__.py` - только include'ы никаких роутеров и view.
** `urls/web.py` - большая часть контента из существующего urls.py
** `urls/api.py` - router для всех viewsets из подпакета api.

Начать предлагаю с list и details actions, чтобы заработало основное ядро. После этого уже будем прикручивать stories по одной.

Что думаешь?

yulai46 commented:
С сериализаторами и урлами все ясно, но пока все равно не до конца понимаю как работает Injector. Завтра попробую все реализовать, думаю тогда и разберусь. 

P.S. Получается я наследуюсь от Injector, ставлю декоратор @model_view_set и работаю с этим как с обычным ModelViewSet от drf ?

yulai46 commented:
@proofit404 Привет! Пытаюсь одолеть ошибку:
`dependencies.exceptions.DependencyError: 'SignUoModelViewSet' can not resolve attribute 'lookup_field'`

**serializers/user.py**
```python
from rest_framework import serializers
from django.contrib.auth.models import User


class UserSerializer(serializers.ModelSerializer):
    password = serializers.CharField(write_only=True)

    class Meta:
        model = User
        fields = ('username', 'email', 'first_name', 'last_name', 'password')
```

**urls/api.py** 
```python
from rest_framework import routers
from example.api.sign_up import SignUoModelViewSet

router = routers.SimpleRouter()
router.register(r'users', SignUoModelViewSet)
urlpatterns = router.urls
```
**api/sign_up.py**
```python
from dependencies.contrib.rest_framework import model_view_set
from dependencies import Injector, Package, operation
from django.contrib.auth.models import User
from rest_framework import status
from rest_framework.response import Response
from example.serializers.user import UserSerializer

implemented = Package("example.implemented")
functions = Package("example.functions")


@model_view_set
class SignUoModelViewSet(Injector):
    queryset = User.objects.all()
    serializer_class = UserSerializer
    sign_up = implemented.SignUp.register_user

    @operation
    def create(request, sign_up):
        serializer = UserSerializer(request.data)
        serializer.is_valid()
        data = serializer.validated_data
        result = sign_up.run(data, request)

        if result.is_success:
            return Response(status=status.HTTP_201_CREATED)
        elif result.failed_on("compare_passwords"):
            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
        elif result.failed_on("validate_password_strength"):
            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
```

proofit404 commented:
Привет. Пока всё в верном направлении.

Приложи полный traceback, пожалуйста. Так будет проще понять где засада.

Сорцы можно коммитить даже неработающие. Так будет проще проводить review.

По мелочи:

```python
serializer = UserSerializer(request.data)
serializer.is_valid()
data = serializer.validated_data
```
В DI scope есть переменная `validated_data`. Лучше использовать её.

Если нет в документации - это мой косяк.

proofit404 commented:
В самом `@model_view_set` из `create` и `update` надо  возвращать инстансы моделей. Из `destroy` не надо возвращать ничего. Если дока косячная, буду очень благодарен за PR.

P.S. Куда-нибудь в директ скинь мне свой адрес. Как этот PR смержим, мы тебе вышлем мерч dry-python.

proofit404 commented:
Any news so far?

proofit404 commented on django/example/api/sign_up.py:
```diff
@@ -0,0 +1,23 @@
+    data = this.validated_data
+
+    @operation
+    def create(request, data, sign_up):
```
We can use `validated_data` as the direct dependency without intermediate alias.

yulai46 replied:
fixed

proofit404 commented on django/example/services/sign_up_user.py:
```diff
@@ -17,6 +17,7 @@ def register_user(I):
         I.persist_profile
         I.login_user
         I.send_welcome_notification
+        I.return_user
```
Maybe `I.show_user` would be a better name. What do you think?

yulai46 replied:
Ok, agree. But why i can't return user object from Success object?

proofit404 replied:
I mean we could rename the method to the more human-friendly language, but keep method body as is.

yulai46 replied:
Understood. I mean a story can return result without calling `Result()` and my question is why i can't get user object from `Success(user=user)` after executing that story

proofit404 replied:
`Result` возвращает значение наружу из story. `Success` только для следующих шагов в сторе.

proofit404 commented on django/example/services/sign_up_user.py:
```diff
@@ -61,6 +62,10 @@ def send_welcome_notification(self, ctx):
         return Success(notification=notification)
 
+    def return_user(self, ctx):
+        print(ctx.user)
```
Please, remove the print function call.

yulai46 replied:
fixed

yulai46 commented:
I have a quetion about authentification.Do I have to use `BasicAuthentification` or `TokenAuthentication`?

proofit404 commented:
In my opinion, it's better to use token authentication or JWT authentication for APIs.

yulai46 commented:
And last question for today. Why we talk in english? )

proofit404 commented:
Потому что я забыл переключиться и ревьювил по умолчанию)).

proofit404 commented:
@yulai46 any news? :pray:

proofit404 commented:
Looks like this is going nowhere.

yulai46 commented:
@proofit404 Простите, на работе завал, думал что смогу совместить но не вышло. Но как будет время я обязательно закончу, если конечно кто то меня не опередит. 
P.S. Зато внедрил dry-python в production :)

proofit404 commented:
Круто! Обязательно напиши feedback по этому поводу)
